### PR TITLE
AA: abstraction for converting KSTypeArgument to KtTypeArgument

### PR DESCRIPTION
**Pull request by neetopia**
unmutes 2 tests along the way.

### CODE REVIEW

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt [hunk 1]
```diff
@@ -77,7 +77,11 @@ class KSClassDeclarationImpl private constructor(private val ktClassOrObjectSymb
     }
 
     override fun asType(typeArguments: List<KSTypeArgument>): KSType {
-        TODO("Not yet implemented")
+        return analyze {
+            analysisSession.buildClassType(ktClassOrObjectSymbol) {
```
Comment (ting-yuan) [inline kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: Is `buildClassType` cached? The type arguments are not cached so it looks like the type isn't cached as well. Do two different instances compare to equal?

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSTypeReferenceImpl.kt [hunk 2]
```diff
@@ -71,7 +72,11 @@ class KSTypeReferenceImpl(private val ktType: KtType) : KSTypeReference {
     }
 
     override val modifiers: Set<Modifier>
-        get() = TODO("Not yet implemented")
+        get() = if (ktType is KtFunctionalType && ktType.isSuspend) {
+            setOf(Modifier.SUSPEND)
```
Comment (ting-yuan) [inline kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSTypeReferenceImpl.kt:76]: Can we leave a todo in the comment to address other possibilities? Looks like it only handled a special case for the test.

Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSTypeReferenceImpl.kt:76]: We only treat `SUSPEND` modifier for type reference in FE1.0 implementation.

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt (hunk 1 above)
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: calling with same argument will result in same return result by reference. But it is not documented behavior, might be subject to change in upstream, let me think about how to cache for this.

Comment (ting-yuan) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: I'm afraid that it is always called with different arguments: `KSTypeArgument.toKtTypeArgument()` always returns new instances.

Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: it's true that the result for `toKtTypeArgument` returns a different instance, but somehow the end result of `buildClassType` can still return same instance, which was my point in the comment, my guess is that there is some underlying caching that checks for this, therefore different instance can still result in same result if they are equal, it does look odd to me tbh.

btw the implementation is not cached in FE1.0 as well, which might be yet a bigger concern for us.

Comment (ting-yuan) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: It's fine as long as either `KtType` is cached or handles `equals` / `hashCode`. Can you confirm that this case is covered in our tests before merging? I.e., calling `asType` twice results in instances, same or not, that compare to equal.

**Review by ting-yuan — APPROVED** [review]

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt (hunk 1 above)
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassDeclarationImpl.kt:81]: after some deep dive, my current understanding of how everything works for both implementations:

* for AA, during type creation, AA first looks up class symbol by `ClassId` which will ensure same instance of type constructor, then for the type arguments, AA will unwrap `KtTypeArgument`s to build FIR type arguments, therefore eliminating instance differences for the inputs. Then for the AA types, its `hashCode` and `equals` functions are overridden to use the FIR's `coneKotlinType`'s, which in turn eliminates all input difference. 

* for FE1.0, types related computations seems to go through a storage manager which is in turn a global cache therefore cached automatically.

**Review by ting-yuan — APPROVED** [review]

**Pull request merged**